### PR TITLE
Printing correct dimension while indexing

### LINKED ISSUE #16192: Dimension error always says dimension 0
## 🐛 Bug

Out of bounds dimension errors while indexing always mention dimension 0, regardless of the dimension being accessed

## To Reproduce

    test_indexing = torch.tensor([[1,2,3],[4,5,6]])
    print(test_indexing[7,0]) 

This gives the message `IndexError: index 7 is out of bounds for dimension 0 with size 2`.

    print(test_indexing[0,7])

This gives the error `IndexError: index 7 is out of bounds for dimension 0 with size 3`

   
## Expected behavior

I assume the second print statement should reference dimension 1, not 0.

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.4.2
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip3] faster-rcnn (0.1, /home/mehdi2277/Documents/HarveyMuddWork/Neural_Nets_Research/temp-pedro-ci/gan-image-metrics/faster-rcnn.pytorch-pytorch-1.0/lib)
[pip3] msgpack-numpy (0.4.3.1)
[pip3] numpy (1.15.4)
[pip3] pytorch-pretrained-bert (0.3.0)
[pip3] torch (1.0.0)
[pip3] torchfold (0.1.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect

### PR DESCRIPTION
**Pull request by bhushan23**
Tensor is visited by slicing one dimension at a time and hence passes dim 0 always.
Using dim() information from object instead of value

 - fixes #16192

### CODE REVIEW

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 1]
```diff
@@ -97,7 +97,7 @@ static Variable applySelect(const Variable& self, int64_t dim, int64_t index) {
   int64_t size = self.size(dim);
   if (index < -size || index >= size) {
     throw IndexError("index %lld is out of bounds for dimension %lld with size %lld",
-      index, dim, size);
+      index, self.dim(), size);
```
Comment (vishwakftw) [inline torch/csrc/autograd/python_variable_indexing.cpp]: I think this is incorrect, `self.dim()` returns the number of dimensions of the tensor IIRC.

Comment (bhushan23) [reply torch/csrc/autograd/python_variable_indexing.cpp]: You are right @vishwakftw 
Instead, full variable needs to be passed from applySlicing and update dimension

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 2]
```diff
@@ -152,7 +152,8 @@ static Variable applySlicing(const Variable& self, PyObject* index, variable_lis
   for (int64_t i = 0; i < size; i++) {
     PyObject* obj = PyTuple_GET_ITEM(index, i); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
     if (THPUtils_checkLong(obj)) {
-      result = applySelect(result, dim, THPUtils_unpackLong(obj));
+      result = applySelect(self, dim, THPUtils_unpackLong(obj));
```
Comment (gchanan) [inline torch/csrc/autograd/python_variable_indexing.cpp]: why did you change the tensor that is passed from `result` to `self`?
